webui/react/src/components/kit/CodeEditor.tsx

```diff
@@ -0,0 +1,367 @@
+} as const;
+
+type Config = ValueOf<typeof Config>;
+
+const descForConfig = {
+  [Config.Submitted]: 'original submitted config',
+  [Config.Runtime]: 'after merge with defaults and templates',
+};
+
+const isConfig = (key: unknown): key is Config =>
+  key === Config.Submitted || key === Config.Runtime;
+
+/**
+ * A component responsible to enable the user to view the code for a experiment.
+ *
+ * It renders a file tree and a selected file in the MonacoEditor
+ *
+ * Props:
+ *
+ * files: an array of one or more files to display code;
+ *
+ * onSelectFile: called with filename when user changes files in the tree;
+ *
+ * readonly: prevent user from making changes to code files;
+ *
+ * selectedFilePath: gives path to the file to set as activeFile;
+ */
+
+const CodeEditor: React.FC<Props> = ({ files, onSelectFile, readonly, selectedFilePath }) => {
+  const [pageError, setPageError] = useState<PageError>(PageError.None);
+
+  const [activeFile, setActiveFile] = useState<TreeNode | null>(files.length > 0 ? files[0] : null);
+  const [downloadInfo, setDownloadInfo] = useState(DEFAULT_DOWNLOAD_INFO);
+  const configDownloadButton = useRef<HTMLAnchorElement>(null);
+  const timeout = useRef<NodeJS.Timeout>();
+
+  const viewMode = useMemo(() => (files.length === 1 ? 'editor' : 'split'), [files.length]);
+  const editorMode = useMemo(() => {
+    const isIpybnFile = /\.ipynb$/i.test(String(activeFile?.key || ''));
+    return isIpybnFile ? 'ipynb' : 'monaco';
+  }, [activeFile]);
+
+  const downloadHandler = useCallback(() => {
+    timeout.current = setTimeout(() => {
+      URL.revokeObjectURL(downloadInfo.url);
+    }, 2000);
+  }, [downloadInfo.url]);
+
+  const fetchFile = useCallback(async (fileInfo: TreeNode) => {
+    if (!fileInfo) return;
+    setPageError(PageError.None);
+
+    if (isConfig(fileInfo.key) || fileInfo.content !== NotLoaded) {
+      setActiveFile(fileInfo);
+      return;
+    }
+
+    let file,
+      content: Loadable<string> = NotLoaded;
+    try {
+      file = await fileInfo.get?.(String(fileInfo.key));
+    } catch (error) {
+      handleError(error, {
+        publicMessage: 'Failed to load selected file.',
+        publicSubject: 'Unable to fetch the selected file.',
+        silent: false,
+        type: ErrorType.Api,
+      });
+      setPageError(PageError.Fetch);
+    }
+
+    try {
+      const text = decodeURIComponent(escape(window.atob(file || '')));
+
+      if (!text) setPageError(PageError.Empty); // Emmits a "Empty file" error message
+      content = Loaded(text);
+    } catch {
+      setPageError(PageError.Decode);
+    }
+    setActiveFile({
+      ...fileInfo,
+      content,
+    });
+  }, []);
+
+  const treeData = useMemo(() => {
+    if (selectedFilePath && activeFile?.key !== selectedFilePath) {
+      const matchTopFileOrFolder = files.find((f) => f.key === selectedFilePath);
+      if (matchTopFileOrFolder) {
+        fetchFile(matchTopFileOrFolder);
+      }
+    }
```

Comment:
Move this to an effect that only runs on mount -- it's causing `fetchFile` to be called multiple times when switching files.

Reply:
I think this is fixed by placing this inside the `useEffect` ?

Reply:
Almost -- given that we already fetch the file on select in the `handleSelectFile` callback, we want to limit this effect to only run on the initial mount, please add an empty dependencies array to this effect.

Reply:
sounds right to me!